docs(gpu-operator): document passthrough variant host-driver prerequisite (#2767)
cozystack/cozystack: docs/gpu-vgpu.md

> ### Symptom

Reviewer: Minor: the Symptom section quotes logs but not the first thing an operator sees in `kubectl get pods` — `nvidia-vfio-manager` in `Init:Error` / `Init:CrashLoopBackOff`. Per the upstream code path the init container exits non-zero (it doesn't complete quietly), so adding one line about the pod state would help, since that's what people grep for first.

  device not bound to 'vfio-pci'; device: 0000:18:00.0 driver: 'nvidia'
```

`lspci -nnk -d 10de:` still shows `Kernel driver in use: nvidia` on every target GPU, the node carries `nvidia.com/gpu.deploy.driver=pre-installed`, and `kubectl get node <NODE> -o json | jq '.status.allocatable | with_entries(select(.key | startswith("nvidia.com/")))'` reports `{}` — no GPU resource was registered.

### Clean-host workaround

Purge the NVIDIA host stack and blacklist the kernel modules so the host never re-claims the GPU. Use an **explicit package list** — `apt autoremove` is dangerous here because the `nvidia-` prefix is shared with NVIDIA DOCA / Mellanox / InfiniBand userspace, and a blanket autoremove can take those out at the same time on a converged GPU + RDMA host:

```bash
# nvidia-dkms-* and nvidia-kernel-* are the load-bearing kernel
# pieces — without them DKMS rebuilds nvidia.ko on the next reboot
# and the blacklist below is bypassed by any explicit modprobe.
# libnvidia-* is intentionally split into the userspace driver
# pieces only — a bare 'libnvidia-*' glob also matches DOCA-OFED
# shipped libnvidia packages on converged GPU + RDMA hosts.

> sudo apt purge \

Reviewer: **The headline remediation can abort with zero effect on a common host shape.** `apt`/`apt-get` treat `*` args as regexes matched against the whole package cache, and a pattern (or a plain name like `cuda-drivers`) that matches **nothing in the cache** is a fatal `E: Unable to locate package` — the entire transaction aborts and *nothing* is purged. `cuda-toolkit-*` / `cuda-drivers` / `cuda-drivers-*` only exist in the cache when NVIDIA's CUDA repo is configured; on a stock Ubuntu 24.04 host with just `nvidia-driver-580-open` from the Ubuntu archive (the very host this doc was verified against), this command fails outright. Note also these are substring regexes, not shell globs — broader than they look, which interacts with the DOCA caution above.

Suggest either a dry-run-first instruction (`sudo apt-get -s purge … | grep -E '^(Purg|Remv)'`, review, drop patterns apt can't locate), or build the list from *installed* packages, which sidesteps both the zero-match abort and the regex over-match:

```bash
dpkg-query -W -f '${Package}\n' 'nvidia-*' 'libnvidia-*' 'cuda-*' 2>/dev/null
```

(dpkg-query patterns are true globs over installed packages.)

  'nvidia-driver-*' 'nvidia-dkms-*' 'nvidia-kernel-*' \
  'nvidia-firmware-*' 'nvidia-utils-*' \
  'nvidia-modprobe' 'nvidia-settings' 'nvidia-prime' \
  'libnvidia-cfg1-*' 'libnvidia-compute-*' 'libnvidia-decode-*' \
  'libnvidia-encode-*' 'libnvidia-extra-*' 'libnvidia-fbc1-*' \
  'libnvidia-gl-*' \
  'cuda-toolkit-*' 'cuda-drivers' 'cuda-drivers-*'

sudo tee /etc/modprobe.d/blacklist-nvidia.conf > /dev/null <<'EOF'
blacklist nouveau
blacklist nvidia
blacklist nvidia_drm
blacklist nvidia_modeset
blacklist nvidia_uvm
EOF

> sudo update-initramfs -u

Reviewer: `update-initramfs -u` only rebuilds the **running** kernel's initramfs. If the node boots a different (e.g. just-upgraded) kernel after this reboot, its initramfs won't carry the blacklist and the host can re-claim the GPU. Use `-k all`:

```suggestion
sudo update-initramfs -u -k all
```

sudo reboot
```

> After the reboot, confirm the host is clean and kick the stuck DaemonSet pods so they re-create against the cleaned state — they have already labeled the node and will not retry on their own:

Reviewer: **Factual error — verified against upstream v0.10.0 source.** `uninstallDriver()` returns `fmt.Errorf("driver is pre-installed on host")` right after the labeling step ([`cmd/driver-manager/main.go#L272-L281`](https://github.com/NVIDIA/k8s-driver-manager/blob/v0.10.0/cmd/driver-manager/main.go#L272-L281)), i.e. the init container exits **non-zero**. kubelet keeps restarting it under crash-loop backoff, so it *does* retry on its own — and the reboot in this workaround recreates the pod sandbox and re-runs init containers anyway. The `kubectl delete` is still worth keeping (it skips the up-to-5-minute backoff wait), but the stated justification is wrong.

```suggestion
After the reboot, confirm the host is clean. The init container exits non-zero after detecting a host driver, so both DaemonSet pods sit in `Init:CrashLoopBackOff`; they *will* retry on their own, but deleting them skips the up-to-5-minute backoff window:
```

```bash
# Should print nothing.
lsmod | grep -E '^(nvidia|nouveau)'

# Self-evident verification: prints "ok: gone" on success.

> test -e /usr/bin/nvidia-smi && echo "STILL PRESENT — purge incomplete" || echo "ok: gone"

Reviewer: `isHostDriver` runs `chroot /host nvidia-smi` resolved via `PATH` inside the chroot, not pinned to `/usr/bin`. So `test -e /usr/bin/nvidia-smi` misses `/usr/local/bin/nvidia-smi` (typical leftover of a `.run`-installer or CUDA-toolkit install), and the apt purge above doesn't cover `.run`-installed drivers at all. Match what the operator actually does:

```suggestion
command -v nvidia-smi && echo "STILL PRESENT — purge incomplete" || echo "ok: gone"
```

Also worth adding `dkms status | grep -i nvidia` (a leftover DKMS module rebuilds `nvidia.ko` on the next kernel update and the blacklist is bypassed by an explicit modprobe), and one sentence: if the driver was installed with NVIDIA's `.run` installer rather than apt, run `nvidia-uninstall` instead of the purge.

# Force a fresh pass of vfio-manager and the sandbox validator.
# The DaemonSet labels are operator-managed and not stable across
# gpu-operator versions, so delete by name pattern — anchored so the
# match is the two DaemonSets and nothing else.
kubectl -n cozy-gpu-operator get pods -o name \
  | grep -E '^pod/(nvidia-vfio-manager|nvidia-sandbox-validator)-' \
  | xargs -r kubectl -n cozy-gpu-operator delete

> ```

Reviewer: **Stale label left behind — verified against upstream source.** `disableContainerizedDriver()` sets `nvidia.com/gpu.deploy.driver=pre-installed`, but the success path `rescheduleGPUOperatorComponents()` only restores the *operand* labels (validator / toolkit / device-plugin / gfd / dcgm / mig / nvsm / sandbox-validator / sandbox-device-plugin / vgpu-device-manager) — `nvidia.com/gpu.deploy.driver` is **not** in that list and is never reset by the binary. So after a successful recovery the node keeps `pre-installed` indefinitely. That's the exact label the Symptom section tells operators to use as evidence, and it would disable the containerized-driver DaemonSet if the node ever switches to the `container` workload (directly relevant once #2766 lands). Suggest adding a cleanup step:

```bash
# The init container set this label and nothing resets it on the success
# path; clear it so it does not mislead future debugging.
kubectl label node <NODE> nvidia.com/gpu.deploy.driver-
```

(Worth a quick runtime check that the gpu-operator controller doesn't reconcile it back/away on its own — see the verification note in the review.)
